Major updates to Obsidian Khoj plugin chat interface and editing features

hjamet opened the pull request:
## Description
This PR introduces significant improvements to the Obsidian Khoj plugin's chat interface and editing capabilities, enhancing the overall user experience and content management functionality.

## Features

### 🔍 Enhanced Communication Mode
I've implemented radio buttons below the chat window for easier communication mode selection. The modes are now displayed as emojis in the conversation for a cleaner interface, replacing the previous text-based system (e.g., /default, /research). I've also documented the search mode functionality in the help command.

#### Screenshots
- Radio buttons for mode selection
- Emoji display in conversations
![Recording 2025-02-11 at 18 56 10](https://github.com/user-attachments/assets/798d15df-ad32-45bd-b03f-581f6093575a)

### 💬 Revamped Message Interaction
I've redesigned the message buttons with improved spacing and color coding for better visual differentiation. The new edit button allows quick message modifications - clicking it removes the conversation up to that point and copies the message to the input field for easy editing or retrying questions.

#### Screenshots
- New message styling and color scheme
![Recording 2025-02-11 at 18 44 48](https://github.com/user-attachments/assets/159ece3d-2d80-4583-a7a8-2ef1f253adcc)
- Edit button functionality
![Recording 2025-02-11 at 18 47 52](https://github.com/user-attachments/assets/82ee7221-bc49-4088-9a98-744ef74d1e58)

### 🤖 Advanced Agent Selection System
I've added a new chat creation button with agent selection capability. Users can now choose from their available agents when starting a new chat. While agents can't be switched mid-conversation to maintain context, users can easily start fresh conversations with different agents.

#### Screenshots
- Agent selection dropdown
![Recording 2025-02-11 at 18 51 27](https://github.com/user-attachments/assets/be4208df-224c-45bf-a5b4-cf0a8068b102)

### 👁️ Real-Time Context Awareness
I've added a button that gives Khoj access to read Obsidian opened tabs. This allows Khoj to read open notes and track changes in real-time, maintaining a history of previous versions to provide more contextual assistance.

#### Screenshots
- Window access toggle
![Recording 2025-02-11 at 18 59 01](https://github.com/user-attachments/assets/b596bfca-f622-41b7-b826-25a8e254d4a2)

### ✏️ Smart Document Editing
Inspired by Cursor IDE's intelligent editing and ChatGPT's Canvas functionality, I've implemented a first version of a content creation system we've been discussing. Using a JSON-based modification system, Khoj can now make precise changes to specific parts of files, with changes previewed in yellow highlighting before application. Modification code blocks are neatly organized in collapsible sections with clear action summaries. While this is just a first step, it's working remarkably well and I have several ideas for expanding this functionality to make Khoj an even more powerful content creation assistant.

#### Screenshots
- JSON modification preview
- Change highlighting system
- Collapsible code blocks
- Accept/cancel controls
![Recording 2025-02-11 at 19 02 32](https://github.com/user-attachments/assets/88826c9e-d0c9-40da-ab78-9976c786aa9e)

sabaimran commented on src/interface/obsidian/src/chat_view.ts:
```diff
@@ -1598,4 +1959,803 @@ export class KhojChatView extends KhojPaneView {
+                        endIndex = content.length;
+                    } else {
+                        if (startIndex !== -1) {
+                            endIndex = content.indexOf(after, startIndex);
```
You'll probably want to look for the first reference of `after` after `startIndex`.

hjamet replied:
In fact, I don't believe it :D After a bunch of tests, it appears that the LLM performs better this way: it typically allows it to replace sentences or paragraphs by mentioning the beginning of the paragraph to be modified rather than the end of the previous paragraph, which is indeed a bit more logical :)

sabaimran reviewed (commented):
Great work, @hjamet !

hjamet commented on src/interface/obsidian/src/main.ts:
Excellent suggestion! I will try to fix that! I was actually planning to slightly modify the format of the Khoj-edit blocks by using XML instead of code blocks. Otherwise, it could lead to parsing errors. As for the radio buttons, I must admit that personally, I find them more pleasant and intuitive, and I was even about to propose reproducing this system on the web interface to be honest ^^ After all, it's just my opinion as a user ;)

sabaimran replied:
Given your feedback, I would still say we should make the radio buttons a dropdown menu, and place it above the chat input area. It would make it two clicks away to change the setting, but this is a good tradeoff for reducing the visual clutter.

It also seems like now there's a UI bug in the row elements in the navigation, where the new chat function is being crowded out when the chat pane is narrow
<img width="479" alt="image" src="https://github.com/user-attachments/assets/8600f435-b010-4505-abbb-281793555c0f" />

To accommodate this, we should move the New Chat button below the nav menu if it's in a narrow window (say, less than 600px width).

One last thing, let's use the Obsidian icon pack, rather than the emojis, for consistency :). See how the [`setIcon`](https://docs.obsidian.md/Reference/TypeScript+API/setIcon) method is being used in the code. You can find valid icon names [here](https://lucide.dev/icons/).

hjamet commented on src/interface/obsidian/src/chat_view.ts:
```diff
@@ -649,15 +849,69 @@ export class KhojChatView extends KhojPaneView {
+                    const allMessages = Array.from(this.contentEl.getElementsByClassName('khoj-chat-message'));
+                    const currentIndex = allMessages.indexOf(messageEl as HTMLElement);
+
+                    // Remove all messages after and including this one
```
Mon mauvais ! Je vais fix ça : c'est un pur oubli :D

- [ ] Todo

hjamet commented on src/interface/obsidian/styles.css:
```diff
@@ -581,6 +582,35 @@ button.chat-action-button:hover {
+button.chat-action-button[title="Copy Message to Clipboard"]:hover {
+    background-color: rgba(0, 122, 255, 0.1);
+    /* Light blue */
+    border-color: rgb(0, 122, 255);
```
Bonne suggestion : j'ignorais l'option ! Je m'en charge

- [ ] Todo

hjamet commented on src/interface/obsidian/src/chat_view.ts:
```diff
@@ -1598,4 +1959,803 @@ export class KhojChatView extends KhojPaneView {
+
+1. Using just a few words to identify long text (notice how "campaign launch" is kept in content):
+\`\`\`khoj-edit
+{
```
Indeed, it is not perfect, but I have used it for the past two weeks and I have never had a problem to be honest :) Originally, I was using regex for more targeting and bulk modifications, but only Claude managed to make something out of it... once every two times. So I switched to this simpler version :) But in any case, only the first identified and modified element (which prevents accidentally modifying other open files that we might not have in view: it limits the damage ^^)

Once again, I plan to slightly modify this format for XML (because currently, it is impossible to correctly parse Khoj's response when its khoj-edit block itself contains a code block!)

hjamet commented on src/interface/obsidian/src/chat_view.ts:
```diff
@@ -1598,4 +1959,803 @@ export class KhojChatView extends KhojPaneView {
+        openFilesContent += "[END OF CURRENT FILES CONTEXT]\n\n";
+        openFilesContent += "[END OF SYSTEM INSTRUCTIONS]\n\n";
+
+        return openFilesContent;
```
Indeed, I looked for a way to edit system instructions directly through the API, but unfortunately, I couldn't find one! I would really like to add this feature, but I still have some issues getting the backend to work properly in development mode, I must admit! Ideally, system instructions should be modified over time to avoid saturating the model's context and to have a more logical interface... But I'm afraid that will be for a V2 :)

hjamet commented on yarn.lock:
My bad, I removed it ! 🦦

hjamet commented:
UX/UI Improvements & Feature Additions:  
- Implemented KhojEdit XML blocks with live preview and error handling  
- Fixed file editing and access bugs   
- Added new "Similar" side panel for direct semantic search  
- Enhanced document similarity viewing interface  
- Modernized overall design while maintaining minimalist aesthetic  
- Improved permanent access to relevant documents during writing  
- Added real-time semantic search in side panel  
- General UI polish and modernization  

Note: There's an ongoing GitGuardian issue whose origin remains unclear and I don't believe I'm the source. Looking for assistance from someone who could help resolve this issue.

debanjum reviewed (commented):
Hey @hjamet, I've reviewed your PR and pushed some improvements/cleanup. The changes look neat and file edit experience is quite fun.

I've moved it to use the more standard SEARCH/REPLACE form (used by [aider](https://aider.chat/docs/more/edit-formats.html) etc.). Also render more granular diff view of the file edit suggestions by Khoj in chat mesages.

I'll have another look at the changes tomorrow. Have a look at my comments and modifications and let me know if you have any concerns or suggestions?

debanjum commented on src/interface/obsidian/src/chat_view.ts:
```diff
@@ -348,7 +476,7 @@ export class KhojChatView extends KhojPaneView {
     }
 
-    generateReference(messageEl: Element, referenceJson: any, index: number) {
+    generateReference(messageEl: Element, referenceJson: any, index: number | string) {
```
Reverted and passed index as is instead of converting it to string where `generateReference` is being called

debanjum reviewed (approved):
I've pushed another set of improvements. Changes look good from my side now! Let me know if you have question, concerns. Otherwise we can merge this PR in a day or two.

debanjum commented:
Hey @hjamet can you put the new content synchronization commit you pushed into a separate PR? I don't want to increase the size of this PR further. I can review your synchronization changes separately but do not want to delay merging this PR.

hjamet commented:
> Hey @hjamet can you put the new content synchronization commit you pushed into a separate PR? I don't want to increase the size of this PR further. I can review your synchronization changes separately but do not want to delay merging this PR.

Indeed, very bad practice on my part. Since it was a significant bug, I wanted to take advantage of the opportunity to fix it in one go ^^ I'll make a separate pull request for the fix :)

debanjum commented:
Excited to get this PR merged!

> Indeed, very bad practice on my part. Since it was a significant bug, I wanted to take advantage of the opportunity to fix it in one go ^^ I'll make a separate pull request for the fix :)

Also thanks for pulling your syncing changes into a separate PR. Keeping changes modular, makes reviews easier, faster and better :)
